sortingHat/sortingHat.py

```diff
@@ -0,0 +1,347 @@
+            if g.name == g_name:
+                prefs.append(g)
+                break
+    
+    if lead == "NO_PREF":
+        lead = Lead.NO_PREF
+    elif lead == "STRONG_LEAD":
+        lead =  Lead.STRONG_LEAD
+    elif lead == "PREFER_LEAD":
+        lead = Lead.PREFER_LEAD
+    elif lead == "STRONG_FOLLOW":
+        lead = Lead.STRONG_FOLLOW
+    elif lead == "PREFER_FOLLOW":
+        lead = Lead.PREFER_FOLLOW
+    
+    student_to_add = Student()
+    student_to_add.name = name
+    student_to_add.identikey = identikey
+    student_to_add.preferences = prefs
+    student_to_add.known_skills = known
+    student_to_add.skills_to_learn = learn
+    student_to_add.desired_leadership = lead
+#    students.append(student_to_add)
+    return student_to_add
+
+#This function combines all variables, but unused in current implementation
+#May be more applicable to a hungarian algorithm approad
+def calcStudentPreference(known, learn, preference):
+    # weight variables. This should be more extensible to account
+    # for multiple possible variables
+    assert (known >= 0)
+    assert (learn >= 0)
+    assert (preference > 0)
+    pref = 5 * 1/preference + LEARN_WEIGHT * learn + \
+            KNOWN_WEIGHT * known
+    pref = pref/(1 + LEARN_WEIGHT + KNOWN_WEIGHT)
+    # return a value normalized between 0 and 1
+    return (pref - 1)/(8 - 1)
+
+##! Use this to make list a group preferences for gale-shapley
+def calcGroupPreference(known, learn):
+    assert (known >= 0)
+    assert (learn >= 0)
+    pref = LEARN_WEIGHT * learn + KNOWN_WEIGHT * known
+    pref = pref/(KNOWN_WEIGHT + LEARN_WEIGHT)
+    return pref/3 # group pref range is 0-3, so normalize between 0 and 1
+
+##! This function registers students with each connected group and adds the
+##! student to the group's preference list with the associated preference score
+def registerUser(student):
+    known_score = 0
+    learn_score = 0
+    for grp in student.preferences:
+        # How to weight group preferences for students?
+        for attr in grp.skills_needed:
+            if attr in student.known_skills and known_score < 10:
+                known_score += 1 
+            if attr in student.skills_to_learn and learn_score < 10:
+                learn_score += 1
+        grp.preferences[student] = calcGroupPreference(known_score, learn_score)
+        
+
+def trimGroups():
+    g_count = int(float(len(students))/OPT_SIZE)
+    min_edges = 10000 
+    while len(groups) > g_count:
+        for group in groups:
+            if len(group.preferences) < min_edges:
+                group_to_drop = group
+        print("Removing %s" % group)
+        groups.remove(group)
+    for student in students:
+        if group in student.preferences:
+            student.preferences.remove(group)
+
+def unpopular():
+    for group in groups:
+        if len(group.preferences) < MIN_SIZE:
+            print(group.name + " is unpopular")
+
+def swapThemBitches(shortGroup, firstPass):
```

Comment:
Oh I like the name though and think we should add a function doc comment as you've done with the others